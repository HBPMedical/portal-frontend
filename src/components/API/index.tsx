export { default as APICore } from './Core';
export { default as APIMining } from './Mining';
export { backendURL, webURL } from './RequestURLS';
