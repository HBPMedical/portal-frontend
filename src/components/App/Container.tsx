import * as React from 'react';
import ReactGA from 'react-ga';
import { Route, Router } from 'react-router-dom';
import { Provider, Subscribe } from 'unstated';

// import UNSTATED from 'unstated-debug';
import {
  APICore,
  APIExperiment,
  APIMining,
  APIModel,
  APIUser,
  webURL
} from '../API'; // as interfaces
import config from '../API/RequestHeaders';
import App, { AppConfig } from '../App/App';
import Splash from '../UI/Splash';
import { history } from '../utils';
import MIPContext from './MIPContext';

// UNSTATED.logStateChanges = process.env.NODE_ENV === "development";

interface State {
  appConfig: AppConfig;
  showTooltip: boolean;
}

class AppContainer extends React.Component<any, State> {
  public state: State = { appConfig: {}, showTooltip: true };
  private apiExperiment = new APIExperiment(config);
  private apiModel = new APIModel(config);
  private apiCore = new APICore(config);
  private apiMining = new APIMining(config);
  private apiUser = new APIUser(config);

  private intervalId: any; // FIXME: NodeJS.Timer | undefined;

  public async componentDidMount(): Promise<
    [void, void, void, void, void, void] | void
  > {
    // Conf written by dockerize
    let appConfig: AppConfig;
    const response = await fetch(`${webURL}/static/config.json`);
    try {
      const config = await response.json();
      appConfig = {
        ...config,
        datacatalogueUrl:
          config.datacatalogueUrl === '0' ? undefined : config.datacatalogueUrl
      };
      this.setState({ appConfig });

      if (appConfig.ga) {
        ReactGA.initialize(appConfig.ga);
      }
    } catch (e) {
      appConfig = {
        instanceName: 'MIP DEV',
        version: 'alpha',
        datacatalogueUrl: undefined
      };

      this.setState({ appConfig });
    }

    await this.apiUser.user();
    if (this.apiUser.state.authenticated) {
      const username =
        this.apiUser.state.user && this.apiUser.state.user.username;

      // Experiments polling and auth by interval
      this.intervalId = setInterval(() => {
        this.apiUser.user().then(() => {
          if (this.apiUser.state.authenticated) {
            this.apiExperiment.all();
          } else {
            clearInterval(this.intervalId);
          }
        });
      }, 10 * 1000);

      return await Promise.all([
        this.apiUser.profile({ username }),
        this.apiExperiment.all(),
        this.apiCore.fetchPathologies(),
        this.apiCore.algorithms(),
        this.apiCore.stats(),
        this.apiModel.all()
      ]);
    }

    return Promise.resolve();
  }

  public componentWillUnmount(): void {
    clearInterval(this.intervalId);
  }

  public render(): JSX.Element {
    const toggleTooltip = (): void => {
      this.setState(state => ({
        ...state,
        showTooltip: !state.showTooltip
      }));
    };

    return (
      <MIPContext.Provider
        value={{ showTooltips: this.state.showTooltip, toggleTooltip }}
      >
        <Provider
          inject={[
            this.apiExperiment,
            this.apiCore,
            this.apiModel,
            this.apiMining,
            this.apiUser
          ]}
        >
          <Router history={history}>
            <Subscribe
              to={[APIExperiment, APICore, APIModel, APIMining, APIUser]}
            >
              {(
                apiExperiment: APIExperiment,
                apiCore: APICore,
                apiModel: APIModel,
                apiMining: APIMining,
                apiUser: APIUser
              ): JSX.Element => {
                const loading = apiUser.state.loading;
                const authenticated = apiUser.state.authenticated;

                return (
                  <>
                    <Route
                      // Callback from the auth server, redirected to the API
                      path="/services/login/hbp"
                      // tslint:disable-next-line jsx-no-lambda
                      render={(props): JSX.Element => {
                        const {
                          location: { search }
                        } = props;
                        apiUser.login(search);

                        return <div />;
                      }}
                    />
                    {!loading && !authenticated && (
                      <Splash
                        forbidden={this.apiUser.state.forbidden}
                        logout={apiUser.logout}
                      />
                    )}
                    {!loading && authenticated && (
                      <App
                        appConfig={this.state.appConfig}
                        apiExperiment={apiExperiment}
                        apiCore={apiCore}
                        apiModel={apiModel}
                        apiMining={apiMining}
                        apiUser={apiUser}
                      />
                    )}
                  </>
                );
              }}
            </Subscribe>
          </Router>
        </Provider>
      </MIPContext.Provider>
    );
  }
}

export default AppContainer;
