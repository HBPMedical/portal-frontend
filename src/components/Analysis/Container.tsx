import '../Model.css';
import './Review.css';

import * as React from 'react';
import { Panel } from 'react-bootstrap';
import { RouteComponentProps } from 'react-router-dom';

import { APICore, APIMining, APIModel } from '../API';
import { VariableEntity } from '../API/Core';
import { MiningPayload } from '../API/Mining';
import { ModelResponse } from '../API/Model';
import { variablesFilter } from '../constants';
import { IAlert } from '../UI/Alert';
import LargeDatasetSelect from '../UI/LargeDatasetSelect';
import Model from '../UI/Model';
import Content from './Content';
import Filter from './Filter';
import ExperimentReviewHeader from './Header';

interface Props extends RouteComponentProps {
  apiModel: APIModel;
  apiCore: APICore;
  apiMining: APIMining;
}
interface State {
  alert?: IAlert;
  loadingSummary?: boolean;
  summaryStatistics?: any;
}

const Container = ({
  apiModel,
  apiCore,
  apiMining,
  ...props
}: Props): JSX.Element => {
  const { history } = props;

  const trainingDatasets =
    apiModel.state.model && apiModel.state.model.query.trainingDatasets;
  const queryfilters =
    apiModel.state.model && apiModel.state.model.query.filters;
  const [shouldReload, setShouldReload] = React.useState(true);

  React.useEffect(() => {
    const query = apiModel.state.model && apiModel.state.model.query;
    const datasets = query && query.trainingDatasets;

    if (!shouldReload) {
      return;
    }

    if (datasets && query) {
      const payload: MiningPayload = {
        covariables: query.coVariables ? query.coVariables : [],
        datasets,
        filters: query.filters ? query.filters : '',
        grouping: query.groupings ? query.groupings : [],
        variables: query.variables ? query.variables : [],
        pathology: query.pathology ? query.pathology : ''
      };

      apiMining.descriptiveStatisticsByDataset({ payload });
    }
  }, [
    apiModel.state.model,
    trainingDatasets,
    queryfilters,
    apiMining,
    shouldReload
  ]);

  const handleSaveModel = async ({
    title
  }: {
    title: string;
  }): Promise<void> => {
    setShouldReload(false);
    const model = apiModel.state.model;
    apiModel.save({ model, title });
  };

  const handleRunExperiment = async (): Promise<void> => {
    const model = apiModel.state.model;
    if (model) {
      apiMining.abortMiningRequests();
      await apiModel.update({ model });
      history.push(`/experiment`);
    }
  };

  const handleGoBackToExplore = (): void => {
    history.push(`/explore`);
  };

  const handleUpdateFilter = async (filters: string): Promise<void> => {
    const model = apiModel.state.model;
    if (model) {
      model.query.filters = (filters && JSON.stringify(filters)) || '';
      setShouldReload(true);
      await apiModel.setModel(model);
    }
  };

  const handleSelectModel = async (model?: ModelResponse): Promise<void> => {
    if (model) {
      setShouldReload(true);
      await apiModel.setModel(model);
    }
  };

  const makeFilters = ({
    apiCore,
    apiModel
  }: {
    apiCore: APICore;
    apiModel: APIModel;
  }): any => {
    const query = apiModel.state.model && apiModel.state.model.query;
    const variables = apiCore.variablesForPathology(query && query.pathology);

    // FIXME: move to Filter, refactor in a pure way
    let fields: any[] = [];
    const buildFilter = (code: string) => {
      if (!variables) {
        return [];
      }

      const originalVar = variables.find(
        (variable: VariableEntity) => variable.code === code
      );

      if (!originalVar) {
        return [];
      }

      const output: any = {
        id: code,
        label: originalVar.label || originalVar.code,
        name: code
      };

      if (originalVar && originalVar.enumerations) {
        output.values = originalVar.enumerations.map((c: any) => ({
          [c.code]: c.label || c.code
        }));
        output.input = 'select';
        output.operators = ['equal', 'not_equal', 'in', 'not_in'];
      }

      const type = originalVar && originalVar.type;
      if (type === 'real') {
        output.type = 'double';
        output.input = 'number';
        output.operators = [
          'equal',
          'not_equal',
          'less',
          'greater',
          'between',
          'not_between'
        ];
      }

      if (type === 'integer') {
        output.type = 'integer';
        output.input = 'number';
        output.operators = [
          'equal',
          'not_equal',
          'less',
          'greater',
          'between',
          'not_between'
        ];
      }

      return output;
    };

    const keys = ['variables', 'coVariables', 'groupings', 'filtersFromParams'];
    let allVariables: string[] = [];
    if (query) {
      keys.forEach((key: string) => {
        const rows = query[key];
        if (rows) {
          rows.forEach((v: any) => {
            allVariables.push(v.code);
          });
        }
      });
    }

    allVariables = [...allVariables, ...variablesFilter];

    // add filter variables
    const extractVariablesFromFilter = (filter: any) =>
      filter.rules.forEach((r: any) => {
        if (r.rules) {
          extractVariablesFromFilter(r);
        }
        if (r.id) {
          allVariables.push(r.id);
        }
      });

    if (query && query.filters) {
      extractVariablesFromFilter(JSON.parse(query.filters));
    }

    const allUniqVariables = Array.from(new Set(allVariables));
    fields =
      (variables &&
        [...allUniqVariables.map(buildFilter)].filter((f: any) => f.id)) ||
      [];
    const filters = (query && query.filters && JSON.parse(query.filters)) || '';

    return { query, filters, fields };
  };

  const { fields, filters } = makeFilters({ apiCore, apiModel });
  const model = apiModel.state.model;
  const query = model?.query;
  const datasets = apiCore.datasetsForPathology(query?.pathology);
  const selectedDatasets = model?.query?.trainingDatasets?.map(d => ({
    ...datasets?.find(v => v.code === d.code),
    ...d
  }));

  return (
    <div className="Model Review">
      <div className="header">
        <ExperimentReviewHeader
          handleGoBackToExplore={handleGoBackToExplore}
          handleSaveModel={handleSaveModel}
          handleRunAnalysis={handleRunExperiment}
          model={model}
        />
      </div>
      <div className="content">
        <div className="sidebar">
          <Panel className="datasets">
            <Panel.Body>
              <h5>
                <strong>Pathology</strong>
              </h5>
              <p>{query?.pathology}</p>
              <h5>
                <strong>Datasets</strong>
              </h5>
              <LargeDatasetSelect
                datasets={apiCore.datasetsForPathology(
                  query && query.pathology
                )}
                handleSelectDataset={apiModel.selectDataset}
                selectedDatasets={query?.trainingDatasets || []}
              ></LargeDatasetSelect>
            </Panel.Body>
          </Panel>
          <Model
            model={model}
            selectedSlug={model && model.slug}
            lookup={apiCore.lookup}
            items={apiModel.state.models}
            handleSelectModel={handleSelectModel}
          />
        </div>
        <div className="results">
          <Content
            apiMining={apiMining}
            model={model}
            selectedDatasets={selectedDatasets}
            lookup={apiCore.lookup}
          >
            <Panel className="filters" defaultExpanded={false}>
              <Panel.Title toggle={true}>
                <h3 className={'btn btn-info'}>Filters</h3>
              </Panel.Title>
              <Panel.Collapse>
                <Panel.Body collapsible={true}>
                  {fields && fields.length > 0 && (
                    <Filter
                      rules={filters}
                      filters={fields}
                      handleChangeFilter={handleUpdateFilter}
                    />
                  )}
                </Panel.Body>
              </Panel.Collapse>
            </Panel>
          </Content>
        </div>
      </div>
    </div>
  );
};

export default Container;
