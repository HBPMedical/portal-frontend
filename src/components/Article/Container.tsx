import React, { useEffect, useState } from 'react';
import { Panel } from 'react-bootstrap';
import { RouteComponentProps } from 'react-router-dom';
import { draftStateToHTML } from 'react-wysiwyg-typescript';
import styled from 'styled-components';

import { APICore } from '../API';
import { Article } from '../API/Core';
import Edit from './Edit';
import Header from './Header';

export enum Mode {
  default,
  editing,
  creating
}

const Layout = styled.div`
  display: flex;
  flex-direction: column;
`;

const Content = styled.div`
  display: flex;
  .panel:first-child {
    flex: 1;
    min-width: 360px;
    margin-right: 8px;
    button {
      white-space: normal;
      text-align: left;
    }
  }
`;

const Submit = styled.div`
  float: right;
`;

const initialFormState: Article = { slug: '', title: '' };

interface Params {
  slug: string;
}
interface Props extends RouteComponentProps<Params> {
  apiCore: APICore;
}

export default ({ apiCore, ...props }: Props) => {
  const [currentArticle, setCurrentArticle] = useState<Article>(
    initialFormState
  );
  const [editing, setEditing] = useState(Mode.default);
  const slug = props.match.params.slug;

  useEffect(() => {
    const articles = apiCore.state && apiCore.state.articles;

    if (!articles) {
      return;
    }

    if (slug !== undefined) {
      if (slug === 'create') {
        setEditing(Mode.creating);
        return;
      }

      const nextArticle = articles.find(a => a.slug === slug);
      if (nextArticle) {
        setCurrentArticle(nextArticle);
      }
    } else if (!currentArticle) {
      setCurrentArticle(articles[0]);
    }
  }, [slug, apiCore.state, currentArticle]);

  const handleSelectArticle = (article: Article) => {
    setEditing(Mode.default);
    setCurrentArticle(article);
  };

  const updateArticle = (id: string, article: Article) => {
    apiCore.updateArticle(id, article).then(() => {
      apiCore.articles();
      setEditing(Mode.default);
    });
  };

  const createArticle = (article: Article) => {
    apiCore.createArticle(article).then(() => {
      apiCore.articles();
      setEditing(Mode.default);
    });
  };

  const handleSaveArticle = (id: string, article: Article) => {
    editing === Mode.editing
      ? updateArticle(article.slug, article)
      : createArticle(article);
  };

  const handleNewArticle = () => {
    setCurrentArticle(initialFormState);
    setEditing(Mode.creating);
  };

  return (
    <Layout>
      <Header handleNewArticle={handleNewArticle} />
      <Content>
        <Panel>
          <Panel.Title>
            <h3>My articles</h3>
          </Panel.Title>
          <Panel.Body>
            {apiCore.state &&
              apiCore.state.articles &&
              apiCore.state.articles.map(a => (
                <div key={a.slug}>
                  <button
                    className="btn btn-link"
                    style={{ textTransform: 'none' }}
                    // tslint:disable-next-line jsx-no-lambda
                    onClick={() => handleSelectArticle(a)}
                  >
                    {a.title}
                  </button>
                </div>
              ))}
          </Panel.Body>
        </Panel>
        <Panel>
          <Panel.Body>
            {editing ? (
              <Edit
                currentArticle={currentArticle}
                handleSaveArticle={handleSaveArticle}
                setEditing={setEditing}
              />
            ) : (
              <>
                <Submit>
                  <button
                    className="btn-info btn"
                    // tslint:disable-next-line
                    onClick={() => setEditing(Mode.editing)}
                  >
                    Edit
                  </button>
                </Submit>
                <h1>{currentArticle && currentArticle.title}</h1>
                <h3>{currentArticle && currentArticle.abstract}</h3>
                <div>
                  {currentArticle &&
                    currentArticle.content &&
                    draftStateToHTML(currentArticle.content)}
                </div>
              </>
            )}
          </Panel.Body>
        </Panel>
      </Content>
    </Layout>
  );
};
