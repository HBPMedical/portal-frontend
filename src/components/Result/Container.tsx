import * as React from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { ExperimentResult, ExperimentResultHeader } from '.';
import { APICore, APIExperiment, APIModel } from '../API';
import { ExperimentResponse } from '../API/Experiment';
import '../Experiment.css';
import Model from '../UI/Model';
import Algorithms from './Algorithms';
import Datasets from '../UI/Datasets';

interface RouteParams {
  uuid: string;
  slug: string;
}

interface Props extends RouteComponentProps<RouteParams> {
  apiExperiment: APIExperiment;
  apiModel: APIModel;
  apiCore: APICore;
}

class Experiment extends React.Component<Props> {
  private intervalId: any;

  public async componentDidMount(): Promise<void> {
    const params = this.urlParams(this.props);
    if (!params) {
      return;
    }
    const { uuid, slug } = params;
    const { apiExperiment, apiModel } = this.props;

    await apiExperiment.one({ uuid });
    if (!apiExperiment.loaded()) {
      this.pollFetchExperiment(uuid);
    }

    return await apiModel.one(slug);
  }

  public async componentDidUpdate(prevProps: Props): Promise<void> {
    const params = this.urlParams(this.props);
    if (!params) {
      return;
    }
    const { uuid, slug } = params;
    const previousParams = this.urlParams(prevProps);
    const previousUUID = previousParams && previousParams.uuid;
    const previousSlug = previousParams && previousParams.slug;

    if (slug !== previousSlug) {
      const { apiModel } = this.props;
      await apiModel.one(slug);
    }

    const { apiExperiment } = this.props;
    if (uuid !== previousUUID) {
      await apiExperiment.one({ uuid });
      if (!apiExperiment.loaded()) {
        this.pollFetchExperiment(uuid);
      }
    } else {
      if (apiExperiment.loaded()) {
        clearInterval(this.intervalId);
      }
    }
  }

  public componentWillUnmount(): void {
    clearInterval(this.intervalId);
  }

  public render(): JSX.Element {
    const { apiExperiment, apiModel, apiCore } = this.props;
    return (
      <div className="Experiment Result">
        {/* <div className="header">
          <ExperimentResultHeader
            experiment={apiExperiment.state.experiment}
            experiments={apiExperiment.state.experiments}
            handleSelectExperiment={this.handleSelectExperiment}
            handleShareExperiment={this.handleShareExperiment}
            handleCreateNewExperiment={this.handleCreateNewExperiment}
          />
        </div>
        <div className="content">
          <div className="sidebar">
            <Datasets model={apiModel.state.model} />
            <Model model={apiModel.state.model} lookup={apiCore.lookup} />
            <Algorithms experiment={apiExperiment.state.experiment} />
          </div> */}
        <div className="results">
          <ExperimentResult experimentState={apiExperiment.state} />
          {/* </div> */}
        </div>
      </div>
    );
  }

  private urlParams = (
    props: Props
  ):
    | {
        uuid: string;
        slug: string;
      }
    | undefined => {
    const { match } = props;
    if (!match) {
      return;
    }
    return match.params;
  };

  private handleSelectExperiment = async (
    experiment: ExperimentResponse
  ): Promise<void> => {
    const { modelSlug: modelDefinitionId, uuid } = experiment;
    const { history, apiExperiment } = this.props;
    history.push(`/experiment/${modelDefinitionId}/${uuid}`);
    await apiExperiment.markAsViewed({ uuid });
    return await apiExperiment.one({ uuid });
  };

  private handleShareExperiment = async (): Promise<void> => {
    const { apiExperiment } = this.props;
    const experiment = apiExperiment.state.experiment;
    const shared = experiment && experiment.shared;
    const params = this.urlParams(this.props);

    if (!params) {
      return;
    }

    const { uuid } = params;
    return shared
      ? await apiExperiment.markAsUnshared({ uuid })
      : await apiExperiment.markAsShared({ uuid });
  };

  private handleCreateNewExperiment = (): void => {
    const { history } = this.props;
    history.push(`/experiment`);
  };

  private pollFetchExperiment = (uuid: string): void => {
    clearInterval(this.intervalId);
    const { apiExperiment } = this.props;
    this.intervalId = setInterval(async () => {
      await apiExperiment.one({ uuid });
      if (apiExperiment.loaded()) {
        clearInterval(this.intervalId);
      }
    }, 10 * 1000);
  };
}

export default withRouter(Experiment);
