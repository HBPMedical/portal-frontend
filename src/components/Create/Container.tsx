import * as React from 'react';
import { Panel, Tab, Tabs } from 'react-bootstrap';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { APICore, APIExperiment, APIModel } from '../API';
import {
  Algorithm,
  AlgorithmParameter,
  AlgorithmParameterRequest
} from '../API/Core';
import { ExperimentPayload, ExperimentResponse } from '../API/Experiment';
import { ModelResponse, Query } from '../API/Model';
import { AppConfig } from '../App/App';
import '../Experiment.css';
import { Alert, IAlert } from '../UI/Alert';
import DatasetsForm from '../UI/DatasetsForm';
import Model from '../UI/Model';
import AvailableAlgorithms from './AvailableAlgorithms';
import ExperimentCreateHeader from './Header';
import Help from './Help';
import Parameters from './Parameters';

interface Props extends RouteComponentProps<any> {
  apiExperiment: APIExperiment;
  apiCore: APICore;
  apiModel: APIModel;
  appConfig: AppConfig;
}

interface State {
  parameters?: AlgorithmParameter[];
  algorithm?: Algorithm;
  alert: IAlert;
}

class Container extends React.Component<Props, State> {
  public state!: State; // TODO: double check init https://mariusschulz.com/blog/typescript-2-7-strict-property-initialization

  public render(): JSX.Element {
    const { apiCore, apiModel, apiExperiment } = this.props;
    const alert = this.state && this.state.alert;
    const query = apiModel.state.model && apiModel.state.model.query;

    return (
      <div className="Experiment">
        <div className="header">
          <ExperimentCreateHeader
            model={apiModel.state.model}
            experiments={apiExperiment.state.experiments}
            method={this.state && this.state.algorithm}
            handleGoBackToReview={this.handleGoBackToReview}
            handleSelectExperiment={this.handleSelectExperiment}
            handleSaveAndRunExperiment={this.handleSaveAndRunExperiment}
          />
        </div>
        <div className="content">
          <div className="sidebar">
            <Panel className="datasets">
              <Panel.Body>
                <DatasetsForm
                  datasets={apiCore.datasetsForPathology(
                    query && query.pathology
                  )}
                  query={query}
                  handleUpdateQuery={this.handleUpdateQuery}
                />
              </Panel.Body>
            </Panel>
            <Model
              model={apiModel.state.model}
              selectedSlug={apiModel.state.model && apiModel.state.model.slug}
              items={apiModel.state.models}
              handleSelectModel={this.handleSelectModel}
              lookup={apiCore.lookup}
            />
          </div>
          <div className="parameters">
            <Panel>
              <Panel.Body>
                {alert && (
                  <Alert
                    message={alert.message}
                    title={alert.title}
                    styled={alert.styled}
                  />
                )}
                <Tabs
                  defaultActiveKey={1}
                  id="uncontrolled-create-experiment-tab"
                >
                  <Tab eventKey={1} title="Algorithm">
                    <Parameters
                      algorithm={this.state && this.state.algorithm}
                      parameters={this.state && this.state.parameters}
                      handleChangeParameters={this.handleChangeParameters}
                      query={apiModel.state.model && apiModel.state.model.query}
                      apiCore={apiCore}
                    />
                  </Tab>
                  <Tab eventKey={2} title="About running experiments">
                    <Help />
                  </Tab>
                </Tabs>
              </Panel.Body>
            </Panel>
          </div>
          <div className="sidebar2">
            <Panel>
              <Panel.Title>
                <h3>Available Algorithms</h3>
              </Panel.Title>
              <Panel.Body>
                <AvailableAlgorithms
                  algorithms={apiCore.state.algorithms}
                  lookup={apiCore.lookup}
                  handleSelectMethod={this.handleSelectAlgorithm}
                  model={apiModel.state.model}
                />
              </Panel.Body>
            </Panel>
          </div>
        </div>
      </div>
    );
  }

  private handleSelectModel = async (model?: ModelResponse): Promise<any> => {
    if (model) {
      const { slug } = model;
      const { apiModel } = this.props;
      if (slug) {
        return await apiModel.one(slug);
      }
    }
  };

  private handleSelectAlgorithm = (algorithm: Algorithm): void => {
    this.setState({
      algorithm: algorithm,
      parameters: algorithm && (algorithm.parameters as AlgorithmParameter[])
    });
  };

  private handleUpdateQuery = (query: Query): void => {
    const { apiModel } = this.props;
    const model = apiModel.state.model;
    if (model) {
      model.query = query;
      apiModel.setModel(model);
    }
  };

  private handleChangeParameters = (parameters: AlgorithmParameter[]) => {
    this.setState({ parameters });
  };

  private handleSelectExperiment = async (
    experiment: ExperimentResponse
  ): Promise<any> => {
    const { modelSlug: modelDefinitionId, uuid } = experiment;
    const { apiExperiment, history } = this.props;
    history.push(`/experiment/${modelDefinitionId}/${uuid}`);

    return await apiExperiment.one({ uuid });
  };

  private handleGoBackToReview = (): void => {
    const { history } = this.props;
    history.push(`/review`);
  };

  private handleSaveAndRunExperiment = async (
    experimentName: string
  ): Promise<any> => {
    const { apiModel, apiExperiment, history } = this.props;

    if (!this.state) {
      this.setState({
        alert: {
          message: 'An error occured, please try again later',
          styled: 'error',
          title: 'Info'
        }
      });
      return;
    }

    const model = apiModel.state.model;
    if (!model) {
      this.setState({
        alert: {
          message: 'An error occured, please choose a model',
          styled: 'error',
          title: 'Info'
        }
      });
      return;
    }
    await apiModel.update({ model });
    if (!model.slug) {
      this.setState({ alert: { message: 'Model was not saved' } });
      return;
    }
    const selectedAlgorithm = this.state && this.state.algorithm;
    if (!selectedAlgorithm) {
      this.setState({ alert: { message: 'Select an algorithm' } });
      return;
    }

    const { parameters } = this.state;
    if (!parameters) {
      this.setState({ alert: { message: 'Select an algorithm' } });
      return;
    }

    const nextParameters: AlgorithmParameterRequest[] = parameters.map(p => {
      let value: string = p.value;
      const query = model && model.query;

      if (query) {
        if (p.label === 'x') {
          let covariablesArray =
            (query.coVariables && query.coVariables.map(v => v.code)) || [];
          covariablesArray = query.groupings
            ? [...covariablesArray, ...query.groupings.map(v => v.code)]
            : covariablesArray;

          if (covariablesArray.length > 0) {
            const design = parameters.find(p => p.name === 'design');
            if (design) {
              value =
                design.value === 'additive'
                  ? covariablesArray.toString().replace(/,/g, '+')
                  : covariablesArray.toString().replace(/,/g, '*');
            } else {
              value = covariablesArray.toString();
            }
          }
        }

        if (p.label === 'y') {
          // TEST_PAIRED
          // TODO: this will be replaced by the formula field and should be removed when it occurs
          const isVector = selectedAlgorithm.name === 'TTEST_PAIRED';
          const varCount = (query.variables && query.variables.length) || 0;
          value = isVector
            ? (query.variables &&
                query.variables // outputs: a1-a2,b1-b2, c1-a1
                  .reduce(
                    (vectors: string, v, i) =>
                      (i + 1) % 2 === 0
                        ? `${vectors}${v.code},`
                        : varCount === i + 1
                        ? `${vectors}${v.code}-${query.variables &&
                            query.variables[0].code}`
                        : `${vectors}${v.code}-`,
                    ''
                  )
                  .replace(/,$/, '')) ||
              ''
            : (query.variables &&
                query.variables.map(v => v.code).toString()) ||
              '';
        }

        if (p.label === 'dataset') {
          value =
            (query.trainingDatasets &&
              query.trainingDatasets.map(v => v.code).toString()) ||
            '';
        }

        if (p.label === 'pathology') {
          value = (query.pathology && query.pathology.toString()) || '';
        }

        if (p.label === 'filter') {
          value = (query.filters && query.filters) || '';
        }
      }

      return {
        name: p.name,
        label: p.label,
        value
      };
    });
    const experiment: ExperimentPayload = {
      algorithms: [
        {
          label: selectedAlgorithm.label,
          name: selectedAlgorithm.name,
          parameters: nextParameters,
          type: selectedAlgorithm.type
        }
      ],
      model: model.slug,
      label: selectedAlgorithm.label,
      name: experimentName
    };

    await apiExperiment.create({ experiment });
    const { experiment: e, error } = apiExperiment.state;

    if (error) {
      this.setState({
        alert: {
          message: `${error}`
        }
      });

      return;
    }

    const uuid = (e && e.uuid) || '';
    history.push(`/experiment/${model && model.slug}/${uuid}`);
  };
}

export default withRouter(Container);
