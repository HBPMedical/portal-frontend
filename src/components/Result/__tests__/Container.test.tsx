import config from '../../API/RequestHeaders';
import Container from '../Container';
import * as React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import * as ReactDOM from 'react-dom';
import { APICore, APIExperiment, APIModel } from '../../API';
import { shallow } from 'enzyme';

jest.mock('request-promise-native');
describe('Test Result components', () => {
  const apiExperiment = new APIExperiment(config);
  const apiModel = new APIModel(config);
  const apiCore = new APICore(config);

  let props;
  let component;

  beforeEach(() => {
    props = {
      apiExperiment,
      apiCore,
      apiModel
    };
    component = (
      <Router>
        <Container {...props} />
      </Router>
    );
  });

  it('Result dom renders correctly', () => {
    const wrapper = shallow(component);
    expect(wrapper).toBeDefined();
  });

  it('Result snapshot renders correctly', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
