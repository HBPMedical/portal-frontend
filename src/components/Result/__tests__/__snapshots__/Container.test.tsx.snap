// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Result components Result snapshot renders correctly 1`] = `
<div
  className="Experiment Result"
>
  <div
    className="header"
  >
    <div
      className="panel panel-default"
    >
      <div
        className="panel-body"
      >
        <div
          className="item text"
        >
          <h3>
            Results of Experiment 
            <strong />
             on
             
            <a
              href="/v3/review/undefined"
              onClick={[Function]}
            />
          </h3>
          <h5
            className="item"
          >
            Created 
             by
             
          </h5>
        </div>
        <div
          className="item"
        >
          <button
            className="btn btn-info"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            SHARE EXPERIMENT
          </button>
        </div>
        <div
          className="item"
        >
          <div
            className="dropdown btn-group btn-group-default"
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              className="dropdown-toggle btn btn-default"
              disabled={false}
              id="experiment-dropdown"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              type="button"
            >
              RELATED EXPERIMENTS
               
              <span
                className="caret"
              />
            </button>
            <ul
              aria-labelledby="experiment-dropdown"
              className="dropdown-menu"
              role="menu"
            >
              <li
                className=""
                role="presentation"
              >
                <a
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="menuitem"
                  tabIndex="-1"
                >
                  <strong>
                    Create New Experiment
                  </strong>
                </a>
              </li>
              <li
                className=""
                role="presentation"
              >
                <a
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="menuitem"
                  tabIndex="-1"
                >
                  ---------------------------
                </a>
              </li>
              <div
                onKeyDown={[Function]}
                onSelect={[Function]}
                style={
                  Object {
                    "margin": "8px",
                  }
                }
              >
                <p>
                  You have no running experiments on this model.
                </p>
              </div>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="content"
  >
    <div
      className="sidebar"
    >
      <div
        className="model panel panel-default"
      >
        <div
          className="panel-title"
        >
          <h3
            className="sc-ifAKCX bhzABg"
          >
            Model 
            <strong />
          </h3>
          <div />
        </div>
        <div
          className="sc-bxivhb lmNfdD panel-body"
        >
          <div
            className="loader"
          >
            <div
              className="css-5ol6ka-wrapper-wrapper"
            >
              <div
                className="css-1wj5fae-style"
              />
              <div
                className="css-1e3cplz-style"
              />
            </div>
            <p>
              loading...
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="results"
    >
      <div
        className="panel panel-default"
      >
        <div
          className="panel-title"
        >
          <h3>
            
          </h3>
        </div>
        <div
          className="sc-htpNat ipMFxB panel-body"
        >
          <div
            className="loading"
          >
            <h3>
              Your experiment is currently running...
            </h3>
            <p>
              Please check back in a few minutes. This page will automatically refresh once your experiment has finished executing.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sidebar2"
    />
  </div>
</div>
`;
